### ISSUE TITLE
Code does not converge to the optimal policy (tensorflow 0.12.1 and Keras 1.1.0)

### ISSUE THREAD

**Issue by Amir-Ramezani** [yanpanlau/DDPG-Keras-Torcs#16]
Hi,

I tested your code after downgrading tensorflow to version 0.12.1 and Keras to version 1.1.0. So there was no error running the program but it does not converge to the optimal policy as you can see in the following result:

('Episode', 4, 'Step', 268993, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.098257193098788914, 'Loss', 0.021010929718613625)
('Episode', 4, 'Step', 268994, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.0065970534669316794, 'Loss', 0.042674191296100616)
('Episode', 4, 'Step', 268995, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.00031545438826979712, 'Loss', 0.024418037384748459)
('Episode', 4, 'Step', 268996, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.019568415147379711, 'Loss', 0.014433514326810837)
('Episode', 4, 'Step', 268997, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.018506277374548623, 'Loss', 0.041520103812217712)
('Episode', 4, 'Step', 268998, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.010767906475774695, 'Loss', 0.015593868680298328)
('Episode', 4, 'Step', 268999, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.002420423072458714, 'Loss', 0.0040719900280237198)
('Episode', 4, 'Step', 269000, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.004262467498310836, 'Loss', 0.037771023809909821)
('Episode', 4, 'Step', 269001, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.026749610641322499, 'Loss', 0.028206927701830864)
('Episode', 4, 'Step', 269002, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.0030027620641139251, 'Loss', 0.015061482787132263)
('Episode', 4, 'Step', 269003, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.081173965529346151, 'Loss', 0.036130554974079132)
('Episode', 4, 'Step', 269004, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.10121600213326937, 'Loss', 0.014408881776034832)
('Episode', 4, 'Step', 269005, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.030823456600634971, 'Loss', 0.0036484464071691036)
('Episode', 4, 'Step', 269006, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.01758625249384254, 'Loss', 0.016869166865944862)
('Episode', 4, 'Step', 269007, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.0040950106730622809, 'Loss', 0.0061152027919888496)
('Episode', 4, 'Step', 269008, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.0015352772668305245, 'Loss', 0.0020759631879627705)
('Episode', 4, 'Step', 269009, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.013906418235719725, 'Loss', 0.013349947519600391)
('Episode', 4, 'Step', 269010, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.00019903168120680918, 'Loss', 0.033619172871112823)
('Episode', 4, 'Step', 269011, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.00017914316922484539, 'Loss', 0.026164039969444275)
('Episode', 4, 'Step', 269012, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.003975938587499582, 'Loss', 0.0075099128298461437)
('Episode', 4, 'Step', 269013, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.0088532675322910981, 'Loss', 0.051988624036312103)
('Episode', 4, 'Step', 269014, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.0022390205788955062, 'Loss', 0.021969024091959)
('Episode', 4, 'Step', 269015, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.033058362379728756, 'Loss', 0.0012654899619519711)
('Episode', 4, 'Step', 269016, 'Action', array([[-1.,  1.,  1.]]), 'Reward', -0.019517203300999583, 'Loss', 0.012628044933080673)
('Episode', 4, 'Step', 269017, 'Action', array([[-1.,  1.,  1.]]), 'Reward', 0.0091726067061909666, 'Loss', 0.032372094690799713)

As you can see actions are saturated. Could you help me to fix this problem?

Thanks,

**Comment by yanpanlau** [comment]
Hi, are you using Keras default weight initialization? Would you try to
initialize the weight : say a gaussian distribution with small sigma value
like 0.001 or 0.01?

**Comment by Amir-Ramezani** [comment]
I am not familiar with your Keras version, can you say which part should I change and how should I change?

**Comment by yanpanlau** [comment]
See the API reference below :

https://keras.io/initializers/

**Comment by quhezheng** [comment]
I run into the same issue and exactly around 25000 steps.
Both accelerator and brake are used at the maximum extent. The car stuck on the ground.

**Comment by CNDOTA** [comment]
I met the same problem which the actions went to upper bound and lower bound and kept constant. It troubles me for months. It seems that the setup is important and my Keras version is 2.0.8 and tensorflow version is 1.1.0.

**Comment by Yuan0803** [comment]
Same as quhezheng, I met the same issue that the accelerator and brake are stucked in 1 and -1 after a few episode. Could the writer tell us why it happens and how to fix it. Than you! @yanpanlau

**Comment by yassineCh** [comment]
hi guys, I am facing the same issue. The sigmoid and tanh of the last actornetwork layer are saturated. the output is then 0, 1 or -1. did anyone solve it ?

**Comment by QiyuZ** [comment]
I implemented this project as my course project and met with same question. Our group came up some ideas to solve it. Please have a look. https://github.com/QiyuZ/auto_car_experience_training

**Comment by dawgster** [comment]
I'm facing the same issues, the actor network saturates when I use the code on my custom environment. Ive tried some normalizing the states and rewards with running averages but that did not help. Interestingly I'm able to run on the openAI inverted pendulum environment.

**Comment by m5823779** [comment]
Hello, I meet same problem, when my action dimension equal to one, the result is good, but when my action dimension is two (the activation function is tanh and sigmoid), the output of actor will saturate. By the way, I use batch normalization only in my actor network.  And my code reference from: https://github.com/floodsung/DDPG
And here is the result what I said:   https://github.com/m5823779/DDPG
Have anyone already solve this problem?

**Comment by dvalenciar** [comment]
same problem, any solution so far?
